---
title: "Background"
code-overflow: wrap
---

## Getting started with RNA-seq data analysis

This workshop aims to get you started with analysing your own RNA-seq data by providing an example workflow. At each of the major steps in the workflow we will examine issues that could confound your analysis and consider how different available methods address these issues. 

Rather than prescribing what methods you should use, this workshop aims to make the user aware of the biological and statistical challenges involved in analysing RNA-seq data and asks how different methods address those challenges. By the end of this workshop you should be familiar with the general workflow of RNA-seq analysis and be ready to ask informed questions about the specific methods you should select going forward. 

The workshop includes a small example dataset which will be used by the workshop Instructor to demonstrate an analysis. Attendees may choose to follow along with the analysis in real time or they may choose to focus on note-taking. The workshop includes short exercises to highlight key messages and reinforce your understanding. 

#### Background knowledge
This workflow assumes you are familiar with transcriptomic and biological concepts. In particular, it assumes you are familar with the concept of sequencing libraries and with the data format (i.e., FASTA and FASTQ files). For more information on these topics, see the [Data Carpentry lesson](https://datacarpentry.github.io/wrangling-genomics/index.html) which describes the FASTQ format (i.e., files that contain reads and quality scores for the bases in each read) and [NGS basics by Illumina](https://sapac.illumina.com/science/technology/next-generation-sequencing/beginners.html), which describes the full process of sample to data, and further details on chosing an [NGS RNA library preparation method](https://sapac.illumina.com/techniques/sequencing/ngs-library-prep/rna.html), which describes different Illumina kits for prepping your RNA sample into a sequenceable library. 



## The RNA-seq workflow overview


- What are the key steps in an RNA-seq data analysis workflow? 

- What are the peculiar features of RNA-seq data, and how do we address these features to make sure our analysis is unbiased and accurate? 

- What methods are available to handle RNA-seq data analysis?

We will assume that the aim of our RNA-seq experiment is to identify a set of genes which are differentially expressed between two or more groups, and to derive biologically meaningful information from our gene list(s). 

Assuming this is the case, there are many methods that can be applied to this type of analysis and this can be overwhelming. Here we will look at what we call the core workflow - the major steps that you will take to go from the raw FASTA files (the output from an RNA-seq experiment) to an informative list of differentially expressed genes. 

### Core steps in a common workflow:
1. Quality assessment of the raw data
2. Adapter removal and cleanup
3. Quantification of gene expression
4. Identifying differentially expressed genes
5. Over-representation analysis


As you will see, for some steps there are multiple tools that take a functionally similar approach (*e.g.,* quality assessment and adapter removal). For other steps, especially for quantification and identification of differentially expressed genes, there are different methodologies available. These methodologies take different routes to address complex biological, computational and statistical questions. A key takeaway from this workshop is that it is not always possible to identify a "best" tool - *you must be willing and able to identify the strengths and weaknesses of available tools and decide, based on your question and experiment, which tool is most appropriate for you*. 

Let's look at some peculiar features of RNA-seq data and identify the challenges these raise for us as data analysts. 

## What are the main challenges or decision points?

### Counting or quantifying

This is not an exhaustive overview of RNA-seq data, but as a reminder, the raw data we receive in the form of FASTA files is a series of reads from *somewhere* in the genome. Each read is an indication, or measurement, of gene expression. We must take these reads, identify where in the genome they originated from and then make a measurement of how frequently we observed those reads. The more reads that are mapped back to a specific location in the genome, the more highly expressed that region (gene) is. 

There are two major approaches to measure gene expression from reads: in the first method reads are aligned to a reference genome and are then 'counted' (e.g., using programs such as [featureCounts](https://subread.sourceforge.net/featureCounts.html) or [HTSeq](https://htseq.readthedocs.io/en/latest/htseqcount.html)), while in the second method reads are 'quantified' without the need to align to a reference (these are sometimes referred to as 'pseudocounts'; example programs include [Salmon](https://combine-lab.github.io/salmon/) and [Kallisto](https://pachterlab.github.io/kallisto/about)). There are arguments for each of these types; one major point of difference is if you are after gene-level resolution (count method) or transcript-level resolution (quantify method), but a full discussion of these and other approaches is beyond the scope of this workshop. We will point you towards some examples for each approach. 

### Library sizes and composition

Due to myriad biological and technical effects, each sample will have differing numbers of total reads (different library sizes). When we attempt to compare gene expression across samples, these different library sizes can cause genes to be over- or under-expressed. Before we can compare samples we must correct for library size differences. 

Additionally, we need to be aware of library composition. If a library has one (or a few) genes that are very highly expressed, they can be thought of as 'taking up' a high proportion of the total library of reads. If we compare this skewed library with another library, some genes may look down-regulated in the skewed library. Here is a great resource on [different ways to normalise counts](https://hbctraining.github.io/Intro-to-DGE/lessons/02_DGE_count_normalization.html), which can account for  library size and/or composition, produced by the [Harvard Chan Bioinformatics Core.](https://github.com/hbctraining/Intro-to-DGE)

### Heteroskedasticity

Heteroskedasticity describes a situation where variance changes with the mean. In RNA-seq data the level of variance associated with each gene is somewhat dependent on the mean expression level of the gene. This is an issue since some statistical approaches (specifically, the Poisson distribution) assume homoskedasticity (that is, that variance is not associated with the mean). 

Two approaches to dealing with heteroskedasticity are common in the literature: transform the data to reduce (or remove) the mean-variance association or use a method that is robust to the mean-variance association ([DESeq2](https://doi.org/10.1186/s13059-014-0550-8) with the Negative Binomial distribution). 


### Gene length

Does gene length matter? You might be aware that if a gene is longer, it is more likely to have reads mapped to it. This is important to be aware of, but when we are identifying differentially expressed genes we are always comparing genes to themselves. Therefore, gene length is not a factor when identifying differentially expressed genes. 

However, gene length does need to be taken into account when we carry out over-representation analysis.


## Introduction to the dataset

The example data in this workshop are from [Lee et al.,](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1000299) and include a set of six *Saccharomyces cerevisiae* samples in two groups: three wild-type samples and three RNA-degradation mutants. Note that these particular mutants are significantly different from the wildtype and we expect large differences between our two sample groups. 

The data is a subset of an RNA-sequencing experiment which used single end sequencing. 

::: {.callout-important collapse="true"}
## Extra for experts
These data come from a 2008 study, sequenced on the Illumina Genome Analyzer, using an EST strategy (expressed sequenced tags). This is an older approach; ESTs were generally shorter reads (36-100bp) than what we typically would sequence now for Illumina transcriptomes (100bp-150bp). While there are some limitations to this approach (mainly in the depth of information you can retrieve, e.g., no strandedness information, less or no isoform resolution), for the purpose of this workshop (i.e., showing the workflow and what tools you can use), these data "work" much the same as more modern sequence data. These data can be found on NCBI's Sequence Read Archive under [BioProject: PRJNA109083](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA109083). 
:::


